astyanax-recipes/src/main/java/com/netflix/astyanax/recipes/storage/ObjectReader.java

```diff
@@ -165,9 +181,15 @@ public void run() {
                         }
                     }
                     finally {
-                        executor.shutdown();
-                        if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS)) {
-                            throw new Exception("Took too long to fetch object: " + objectName);
+                        if (!isExecutorServiceProvided) {
+                            executor.shutdown();
+                            try {
+                                if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS))
+                                    throw new Exception("Took too long to fetch object: " + objectName);
+                            } finally {
+                                if (!executor.isTerminated())
+                                    executor.shutdownNow();
```

Comment:
I think this call merits some diagnostics - logging what it returns and whether or not the tasks have actually terminated. Here is what shutdownNow() documentation says:

"This method does not wait for actively executing tasks to terminate. Use awaitTermination to do that.

There are no guarantees beyond best-effort attempts to stop processing actively executing tasks. For example, typical implementations will cancel via Thread.interrupt(), so any task that fails to respond to interrupts may never terminate."

I know I was the one who suggested it, but it seems like that even with this change (which I believe should help a lot), there is still potential for getting into thread count explosion. I think it would be good to at least wait for a few seconds here and log whether or not threads have terminated.

(Completely agree with others on inappropriateness of shutting down a shared executor - it could be used in parallel by multiple ObjectReaders with clearly unexpected results.)